### ISSUE TITLE
Conda package for pypdfium2?

### ISSUE THREAD

**Issue by mara004** [AnacondaRecipes/pypdfium2-feedstock#1]
Hi @boldorider4,

I'm pypdfium2's current maintainer, and incidentally found this repository via GH search.
It looks like this is an attempt to create conda packages for pypdfium2?
Please let me know if I can assist. I don't use conda, but if you have any questions regarding pypdfium2 and its setup code, feel free to ask.

FYI, there has been some past discussion and unsuccessful attempts on conda packaging of pypdfium2.
See https://github.com/pypdfium2-team/pypdfium2/discussions/77, https://github.com/mindee/doctr/issues/113 and https://github.com/conda/conda-build/issues/2611#issuecomment-1239338538.

TLDR: My belief is that this is not elegantly possible given the current state of conda. AFAIK, conda does not support building platform specific packages independent of python version. Apart from that, I couldn't get past the problem that conda-build sets up an isolated environment for each package, which makes it comparatively slow.
Contrast all this to python wheels, where we can craft all packages quickly on a single host.

Kind regards,
@mara004

(CC @frgfm and @felixT2K)

**Comment by mara004** [comment]
I saw you also pulled in https://github.com/AnacondaRecipes/ctypesgen-feedstock.
Note that pypdfium2 uses a [custom fork of ctypesgen](https://github.com/pypdfium2-team/ctypesgen/tree/pypdfium2) ATM which will result in cleaner bindings (especially regarding string handling). Building with mainline ctypesgen is still possible but not tested / officially supported. If I get through some PRs upstream we might eventually switch back to mainline, though.

**Comment by mara004** [comment]
Linking some comments for the record:
* https://github.com/AnacondaRecipes/pypdfium2-feedstock/commit/0a86f2bb0c94730996c2ba11fbf1cdfb896aee10#r124057076 (licensing info is currently incorrect)
* https://github.com/AnacondaRecipes/pypdfium2-feedstock/commit/cd045f19c634da481a290e8b1226706b2d5d92d4#r124058597 (pypdfium2 is not noarch)
* https://github.com/AnacondaRecipes/pypdfium2-feedstock/commit/cd045f19c634da481a290e8b1226706b2d5d92d4#r124059335 and https://github.com/AnacondaRecipes/pypdfium2-feedstock/commit/cd045f19c634da481a290e8b1226706b2d5d92d4#r124059442 (manual versioning is not nice)

**Comment by boldorider4** [comment]
Wow, thanks very much @mara004!

About python: we do not build python independent packages, but we build them for each supported python version (`3.8..3.11`).
I still have to figure out how easy this is to build. So far I haven't run into any blockers, but I'm still working on the dependencies (`ctypesgen`).

About `noarch`: our building practices dictate building for each platform individually, to better diversify support for the different platforms.

About `ctypesgen`: I noticed the dependency on the fork, that's why I opened a [PR](https://github.com/AnacondaRecipes/ctypesgen-pypdfium2-team-feedstock/pull/1) for [ctypesgen-pypdfium2-team-feedstock](https://github.com/AnacondaRecipes/ctypesgen-pypdfium2-team-feedstock).

I'll let you know if I get stuck.

**Comment by boldorider4** [comment]
If you agree, I will close this issue, since we discussed everything. Unless you have more points to raise.

**Comment by mara004** [comment]
Ok.
How will this project go on? Provided there are no problems, would you please notify me once this is ready for the first release, so I can take a look? Thanks!

**Comment by boldorider4** [comment]
We are working on this for an Anaconda customer project. We aren't currently planning on releasing on main (defaults), but if that changes, we're more than happy to let you know.

**Comment by mara004** [comment]
I see. Does this mean the packages are private, or would they be released to a public non-main repo?

**Comment by boldorider4** [comment]
exactly.

**Comment by mara004** [comment]
>>  I see. Does this mean the packages are private, or would they be released to a public non-main repo?

> exactly.

If a question mentions two possibilities, then "exactly" is a bit ambiguous. 😅
But I suppose you mean private, right? If so, why? Of course you don't have to, but if the recipe is open-source already, then why not contribute the built packages to the open-source community as well?

**Comment by boldorider4** [comment]
Sorry about me TLDR'ing. The "exactly" was referred to us making the package private. It is a customer project and we cannot disclose more details. However, should we decide to release the package publicly we will notify the maintainer of pypdfium2 (supposedly, you only).

**Comment by mara004** [comment]
OK, acknowledged.
If something changes, you could simply leave a note on https://github.com/mindee/doctr/issues/113. Thanks.

**Comment by mara004** [comment]
Do you have any (legal) objections if I nevertheless try to integrate conda packaging into pypdfium2 itself?
Just to confirm, the recipe may be reused under `BSD-3-Clause` license, right?

**Comment by mara004** [comment]
@boldorider4 Oh, and I think ctypesgen-pypdfium2-team is built from the wrong branch. `master` is identical to upstream, you'll want the `pypdfium2` branch.

**Comment by boldorider4** [comment]
Oh you're right, I just noticed it in the `pyproject.toml`.
I'm not sure how you would integrate the conda packaging in your project. I'm basically trying to reverse engineer the integration of the dependency on `pdium2` from your project and make a clean recipe out of it. The only way to do this cleanly is by writing an actual recipe.
Anyhow, you're welcome to try and integrate it into your code base but I don't think you should do packaging the way you're currently doing it, meaning by delegating everything to a customized `setupsrc`. We have conda for that ;)

**Comment by mara004** [comment]
> I'm not sure how you would integrate the conda packaging in your project.

Actually I did work on this over the last two days and it's almost done, see the [conda](https://github.com/pypdfium2-team/pypdfium2/compare/main...conda) branch. (Only the CI workflow still needs to be written, but this should be relatively straightforward.)

But it's no secret that it's not nice. Building all the conda packages takes over 15min. Contrast this to PyPI packages, which take only 20s to build, and with much less trouble.

> I'm basically trying to reverse engineer the integration of the dependency on pdium2 from your project and make a clean recipe out of it.

I'm not sure if I understand correctly. You mean the integration of the pdfium dependency in pypdfium2, right?

> The only way to do this cleanly is by writing an actual recipe.

Sure, we need a recipe, the quest was just how to integrate the binaries. In particular, what I want(ed) is
* Ability to build for all platforms on a single host. This I have achieved with workarounds. (The bundled `.pyc` files will be incorrect, but IIRC python should just re-generate them if they don't match.)
* Ability to be platform-specific but python independent. To my knowledge, unfortunately not possible with conda. Therefore we now build for all python versions using conda's `--variants` option, but when covering python 3.8 to 3.11 this results in a 4x slowdown and increase in upload size.

but ...

> but I don't think you should do packaging the way you're currently doing it

I think you're hinting that pdfium should be packaged separately, so that pypdfium2 can just depend on it and be cleanly noarch without any packaging complications, right? That's a great idea.
It was off my mind because it's common practice to bundle binaries in wheels on PyPI, but this could indeed be the solution for the trouble.

I was wondering how to tell the bindings where to find the library, but this should really be doable by passing ctypesgen the relevant libdirs on bindings creation time.
Also I would long have liked an option for a configurable pdfium load path via a python file, so it can be defined both before or on import time. (BTW, ctypesgen's library loader is questionable too and needs revisiting, see https://github.com/ctypesgen/ctypesgen/issues/176)

Sorry for the long text, just wrote down my thinking process. 😅

**Comment by mara004** [comment]
> I'm basically trying to reverse engineer the integration of the dependency on `pdium2` from your project and make a clean recipe out of it. The only way to do this cleanly is by writing an actual recipe.
> Anyhow, you're welcome to try and integrate it into your code base but I don't think you should do packaging the way you're currently doing it, meaning by delegating everything to a customized setupsrc. We have conda for that ;)

TLDR https://github.com/mindee/doctr/issues/113#issuecomment-1686879282: Ah, I finally understand, the solution is to package pdfium separately and make pypdfium2 noarch. In retrospect it's really clear what you meant, I just found the formulation a bit confusing so I didn't grasp it at first.

(PS the C lib is called `pdfium`, not `pdium2`)

**Comment by mara004** [comment]
> Anyhow, you're welcome to try and integrate it into your code base but I don't think you should do packaging the way you're currently doing it, meaning by delegating everything to a customized setupsrc. We have conda for that ;)

For the record, I think what we're doing is fine for PyPI, but not for conda, yes.
Since PyPI is python only but conda is a generic package manager.
